pandas/core/internals.py

```diff
@@ -4444,18 +4635,18 @@ def _lcd_dtype(l):
     have_dt64 = len(counts[DatetimeBlock]) > 0
     have_dt64_tz = len(counts[DatetimeTZBlock]) > 0
     have_td64 = len(counts[TimeDeltaBlock]) > 0
+    have_period = len(counts[PeriodBlock]) > 0
     have_cat = len(counts[CategoricalBlock]) > 0
     # TODO: have_sparse is not used
     have_sparse = len(counts[SparseBlock]) > 0  # noqa
     have_numeric = have_float or have_complex or have_int
 
     has_non_numeric = have_dt64 or have_dt64_tz or have_td64 or have_cat
+    has_datetime_like = have_dt64 or have_dt64_tz or have_td64 or have_period
 
     if (have_object or
-        (have_bool and
-         (have_numeric or have_dt64 or have_dt64_tz or have_td64)) or
-        (have_numeric and has_non_numeric) or have_cat or have_dt64 or
-            have_dt64_tz or have_td64):
+       (have_bool and (have_numeric or has_datetime_like)) or
+       (have_numeric and has_non_numeric) or have_cat or has_datetime_like):
```

Comment:
should prob move this entire mess to ``pandas.types.cast`` (maybe sep PR and do before?)